pkg/cloud/services/ec2/dedicatedhosts.go

```diff
@@ -0,0 +1,260 @@
+/*
+Copyright 2025 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ec2
+
+import (
+	"context"
+	"math"
+	"strings"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/service/ec2"
+	"github.com/aws/aws-sdk-go-v2/service/ec2/types"
+	"github.com/pkg/errors"
+
+	infrav1 "sigs.k8s.io/cluster-api-provider-aws/v2/api/v1beta2"
+	"sigs.k8s.io/cluster-api-provider-aws/v2/pkg/record"
+)
+
+// AllocateDedicatedHost allocates a new dedicated host based on the specification.
+func (s *Service) AllocateDedicatedHost(ctx context.Context, spec *infrav1.DynamicHostAllocationSpec, availabilityZone string) (string, error) {
+	s.scope.Debug("Allocating dedicated host", "instanceFamily", spec.InstanceFamily, "availabilityZone", availabilityZone)
+
+	// Determine the instance type for the dedicated host
+	instanceType := s.determineInstanceType(spec)
+
+	// Set default quantity if not specified
+	quantity := int32(1)
+	if spec.Quantity != nil {
+		quantity = *spec.Quantity
+	}
+
+	// Use the specified AZ or fall back to the provided one
+	targetAZ := availabilityZone
+	if spec.AvailabilityZone != nil {
+		targetAZ = *spec.AvailabilityZone
+	}
+
+	input := &ec2.AllocateHostsInput{
+		InstanceFamily:   aws.String(spec.InstanceFamily),
+		AvailabilityZone: aws.String(targetAZ),
+		Quantity:         aws.Int32(quantity),
+	}
+
+	// Set instance type if specified
+	if instanceType != "" {
+		input.InstanceType = aws.String(instanceType)
+	}
+
+	// Add tags if specified
+	if len(spec.Tags) > 0 {
+		var tagSpecs []types.TagSpecification
+		var tags []types.Tag
+		for key, value := range spec.Tags {
+			tags = append(tags, types.Tag{
+				Key:   aws.String(key),
+				Value: aws.String(value),
+			})
+		}
+		tagSpecs = append(tagSpecs, types.TagSpecification{
+			ResourceType: types.ResourceTypeDedicatedHost,
+			Tags:         tags,
+		})
+		input.TagSpecifications = tagSpecs
+	}
+
+	output, err := s.EC2Client.AllocateHosts(ctx, input)
+	if err != nil {
+		record.Warnf(s.scope.InfraCluster(), "FailedAllocateDedicatedHost", "Failed to allocate dedicated host: %v", err)
+		return "", errors.Wrap(err, "failed to allocate dedicated host")
+	}
+
+	if len(output.HostIds) == 0 {
```

Comment:
does aws give guarantees to always return an id immediately in the output response to the creation?
Otherwise that scenario would need to be handled with an async mechanism to fetch on going allocations